File: packages/search-client/src/transport.ts
/**
 * Low-level HTTP transport for the AACSearch SDK.
 */
import { SdkError, type SdkErrorCode } from "./types";

export function buildHeaders(apiKey: string): Record<string, string> {
	return {
		"Content-Type": "application/json",
		Authorization: `Bearer ${apiKey}`,
	};
}

export async function request<T>(
	baseUrl: string,
	apiKey: string,
	method: string,
	path: string,
	body?: unknown,
	fetchImpl?: typeof fetch,
): Promise<T> {
	const url = `${baseUrl.replace(/\/+$/, "")}${path}`;
	const doFetch: typeof fetch = fetchImpl ?? fetch;

	let response: Response;
	try {
		response = await doFetch(url, {
			method,
			headers: buildHeaders(apiKey),
			body: body !== undefined ? JSON.stringify(body) : undefined,
		});
	} catch (cause) {
		throw new SdkError(
			"network_error",
			0,
			cause instanceof Error ? cause.message : "Network request failed",
		);
	}

	const text = await response.text();
	let payload: unknown = undefined;
	if (text) {
		try {
			payload = JSON.parse(text);
		} catch {
			// non-JSON response — fall through
		}
	}

	if (!response.ok) {
		const code: SdkErrorCode = (payload as { error?: SdkErrorCode })?.error ?? "unexpected";
		const details = (payload as { details?: unknown })?.details;
		throw new SdkError(
			code,
			response.status,
			`${response.status} ${code}${details ? ` — ${JSON.stringify(details)}` : ""}`,
			details,
		);
	}

	return payload as T;
}

File: packages/search-client-python/src/aacsearch/_transport.py
"""Internal HTTP transport for the AACSearch SDK."""

from __future__ import annotations

from typing import Any

import httpx

from ._exceptions import _raise_for_status

DEFAULT_TIMEOUT = 30.0  # seconds


def _request(
    base_url: str,
    api_key: str,
    method: str,
    path: str,
    body: Any = None,
    params: dict[str, str] | None = None,
    timeout: float = DEFAULT_TIMEOUT,
) -> Any:
    """Make an authenticated HTTP request to the AACSearch v1 REST API."""
    url = f"{base_url.rstrip('/')}{path}"

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
        "User-Agent": "aacsearch-python-sdk/0.1.0",
    }

    json_data: Any = None
    if body is not None:
        json_data = body

    with httpx.Client(timeout=timeout) as client:
        response = client.request(
            method=method,
            url=url,
            headers=headers,
            json=json_data,
            params=params,
        )

    _raise_for_status(response)

    if response.status_code == 204:
        return None

    return response.json()
